lib/cretonne/meta/cdsl/ti.py

```diff
@@ -0,0 +1,556 @@
+
+    def add_constraint(self, tv1, tv2):
+        # type: (TypeVar, TypeVar) -> None
+        """
+        Add a new equivalence constraint between tv1 and tv2
+        """
+        self.constraints.append((tv1, tv2))
+
+    def get_uid(self):
+        # type: () -> str
+        r = str(self.idx)
+        self.idx += 1
+        return r
+
+    def __repr__(self):
+        # type: () -> str
+        return self.dot()
+
+    def rank(self, tv):
+        # type: (TypeVar) -> int
+        """
+        Get the rank of tv in the partial order. TVs directly associated with a
+        Var get their rank from the Var (see register()).
+        Internally generated non-derived TVs implicitly get the lowest rank (0)
+        Internal derived variables get the highest rank.
+        """
+        default_rank = 5 if tv.is_derived else 0
+        return self.ranks.get(tv, default_rank)
+
+    def register(self, v):
+        # type: (Var) -> None
+        """
+        Register a new Var v.  This computes a rank for the associated TypeVar
+        for v, which is used to impose a partial order on type variables.
+        """
+        self.vars.add(v)
+
+        if v.is_input():
+            r = 4
+        elif v.is_intermediate():
+            r = 3
+        elif v.is_output():
+            r = 2
+        else:
+            assert(v.is_temp())
+            r = 1
+
+        self.ranks[v.get_typevar()] = r
+
+    def free_typevars(self):
+        # type: () -> Set[TypeVar]
+        """
+        Get the free typevars in the current type env.
+        """
+        tvs = set([self[tv].free_typevar() for tv in self.type_map.keys()])
+        # Filter out None here due to singleton type vars
+        return set(filter(lambda x: x is not None, tvs))
+
+    def normalize(self):
+        # type: () -> None
+        """
+        Normalize by:
+            - collapsing any roots that don't correspond to a concrete TV AND
+              have a single TV derived from them or equivalent to them
+
+        E.g. if we have a root of the tree that looks like:
+
+          typeof_a   typeof_b
+                 \  /
+              typeof_x
+                  |
+                half_width(1)
+                  |
+                  1
+
+        we want to collapse the linear path between 1 and typeof_x. The
+        resulting graph is:
+
+          typeof_a   typeof_b
+                 \  /
+              typeof_x
+        """
+        source_tvs = set([v.get_typevar() for v in self.vars])
+        children = {}  # type: Dict[TypeVar, Set[TypeVar]]
+        for v in self.type_map.values():
+            if not v.is_derived:
+                continue
+
+            t = v.free_typevar()
+            s = children.get(t, set())
+            s.add(v)
+            children[t] = s
+
+        for (a, b) in self.type_map.items():
+            s = children.get(b, set())
+            s.add(a)
+            children[b] = s
+
+        for r in list(self.free_typevars()):
```

Comment:
Here, for example. It is not obvious that this algorithm would yield the same result for all set orderings. (I'm not saying it doesn't).

It would be good to either add a comment describing why the algorithm is order-independent, or iterate over a canonical order of the type variables.